### PR TITLE
Adds Https Redirection and Hsts Middlewares

### PR DESCRIPTION
**Pull request by jkotalik**
Implementation of #258.
Final items:
- Verify which browsers respect the Strict-Transport-Security and check behavior on browser.
- Naming
- Any design/ usability changes

cc @danroth27 @shirhatti @muratg

### CODE REVIEW

**Comment by jkotalik** [comment]
Per offline feedback, 
~~HttpsEnforcement~~ -> HttpsPolicy for everything?

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    public class HstsOptions
+    {
+        /// <summary>
+        /// Sets the max-age parameter of the Strict-Transport-Security header.
+        /// </summary>
+        /// <remarks>
+        /// Max-age is required; defaults to 0.
+        /// See: https://tools.ietf.org/html/rfc6797#section-6.1.1
+        /// </remarks>
+        public int MaxAge { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: TimeSpan.
Also, 0 means to disable HSTS, so it can't be the default.
```
   NOTE:  A max-age value of zero (i.e., "max-age=0") signals the UA to
          cease regarding the host as a Known HSTS Host, including the
          includeSubDomains directive (if asserted for that HSTS Host).
```

Comment (Tratcher) [reply src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: Though https://tools.ietf.org/html/rfc6797#section-11.2 suggests 0 as the default and making users set it so that it's not using an arbitrary value.
If we require users to set it then we should make it nullable and throw if it's not set. If not then we need to come up with a non-zero default. Talk to @blowdart.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: https://tools.ietf.org/html/rfc6797#section-11.2 defines an alternative to a fixed second count where you would define a fixed date and then Max-Age would reflect a count-down to that date. Not sure if that's in wide use or worth implementing. Do other providers offer this variation?

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        /// <summary>
+        /// Invoke the middleware.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <returns></returns>
+        public async Task Invoke(HttpContext context)
+        {
+            context.Response.OnStarting(() =>
+            {
+                if (context.Request.Scheme == "https")
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HstsMiddleware.cs]: Request.IsHttps

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        /// <remarks>
+        /// See: https://tools.ietf.org/html/rfc6797#section-6.1.2
+        /// </remarks>
+        public bool IncludeSubDomains { get; set; }
+
+        /// <summary>
+        /// Sets the preload parameter of the Strict-Transport-Security header.
+        /// </summary>
+        /// <remarks>
+        /// Preload is not part of the RFC specification, but is supported by web browsers
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: remove the Chrome reference, just say some web browsers.

File: src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Microsoft.AspNetCore.HttpsEnforcement
+{
+    /// <summary>
+    /// Extension methods for the HttpsEnforcement middleware.
+    /// </summary>
+    public static class HttpsEnforcementExtensions
+    {
+        /// <summary>
+        /// Adds middleware for enforcing HTTPS for all HttpRequests.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementExtensions.cs]: Including redirecting from HTTP to HTTPS and adding the HSTS header.

File: src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementOptions.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+        /// <summary>
+        ///  Options for using HSTS
+        /// </summary>
+        public HstsOptions HstsOptions { get; set; } // TODO should this be initialized at all? We could remove the bool if so.
+
+        /// <summary>
+        /// The status code to be used for Url Redirection
+        /// </summary>
+        public int StatusCode { get; set; } = 301; // TODO throw ArgumentOutOfRangeException from UrlRewrite redirect rule?
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementOptions.cs]: Use the constant

File: src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementOptions.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        /// </summary>
+        public HttpsEnforcementOptions()
+        {
+            HstsOptions = new HstsOptions();
+        }
+
+        /// <summary>
+        /// Whether to use HTTP Strict-Transport-Security (HSTS) on all HTTPS requests.
+        /// </summary>
+        public bool EnforceHsts { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementOptions.cs]: SendHsts? SetHsts? The HSTS enforcement happens on the client.

File: samples/HttpsSample/Startup.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+        public void ConfigureServices(IServiceCollection services)
+        {
+        }
+
+        public void Configure(IApplicationBuilder app)
+        {
+            var httpsEnforcementOptions = new HttpsEnforcementOptions();
```
[the remark this answers is not in the document]
Comment (javiercn) [reply samples/HttpsSample/Startup.cs]: We should have a helper method for this.
```csharp
services.AddHttpsPolicy(o => o. ...);
```

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        /// </summary>
+        /// <param name="app">The <see cref="IApplicationBuilder"/> instance this method extends.</param>
+        public static IApplicationBuilder UseHsts(this IApplicationBuilder app)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+
+            return app.UseHsts(new HstsOptions());
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HstsExtensions.cs]: HstsOptions here should come from DI.

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsMiddleware.cs [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HttpsEnforcement
+{
+    /// <summary>
+    /// Enables Http Strict Transport Security (HSTS)
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HstsMiddleware.cs]: nit: `HTTP` (casing)

File: src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        /// </summary>
+        /// <param name="app">The <see cref="IApplicationBuilder"/> instance this method extends.</param>
+        /// <returns>The <see cref="IApplicationBuilder"/> with HttpsEnforcement.</returns>
+        public static IApplicationBuilder UseHttpsEnforcement(this IApplicationBuilder app)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+            var httpsEnforcementOptions = new HttpsEnforcementOptions();
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.HttpsEnforcement/HttpsEnforcementExtensions.cs]: This should come from DI

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: I'm in favor of having it nullable and throw if it isn't set. The fixed second count is odd.

Comment (Tratcher) [reply src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: Ask the PMs.

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyBuilderExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        /// <returns>The <see cref="IApplicationBuilder"/> with HttpsPolicy.</returns>
+        public static IApplicationBuilder UseHttpsPolicy(this IApplicationBuilder app)
+        {
+            if (app == null)
+            {
+                throw new ArgumentNullException(nameof(app));
+            }
+            var httpsEnforcementOptions = new HttpsPolicyOptions();
+
+            return app.UseHttpsPolicy(httpsEnforcementOptions);
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyBuilderExtensions.cs]: nit: Policy

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+        /// <summary>
+        ///  Options for using HSTS
+        /// </summary>
+        public HstsOptions HstsOptions { get; set; } // TODO should this be initialized at all? We could remove the bool if so.
+
+        /// <summary>
+        /// The status code to be used for Url Redirection
+        /// </summary>
+        public int StatusCode { get; set; } = StatusCodes.Status301MovedPermanently; // TODO throw ArgumentOutOfRangeException from UrlRewrite redirect rule?
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs]: It's ok as it is. I don't think we need to over protect against a non redirect status code. I would change the property name and the doc comment to reflect that is a redirect status code. Property would be RedirectStatusCode. What do you think?

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+        /// <summary>
+        /// Whether to use HTTP Strict-Transport-Security (HSTS) on all HTTPS requests.
+        /// </summary>
+        public bool SetHsts { get; set; }
+
+        /// <summary>
+        ///  Options for using HSTS
+        /// </summary>
+        public HstsOptions HstsOptions { get; set; } // TODO should this be initialized at all? We could remove the bool if so.
```
Comment (javiercn) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs]: I'm happy with this being initialized by default (to whatever defaults we choose) and for it being null to have the meaning of SetHsts = false. That way there can't be invalid combinations (like the user initializes the setting and forgets to set the flag or viceversa). I also think we follow this pattern in other places. @Tratcher do you have anything against my proposal?

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyBuilderExtensions.cs [hunk 14]
```diff
@@ -65,5 +46,9 @@ public static IApplicationBuilder UseHttpsPolicy(this IApplicationBuilder app, H
 
             return app;
         }
+
+        public static IApplicationBuilder UseHttpsPolicy(this IApplicationBuilder app, bool enableHsts)
+        {
```
Comment (muratg) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyBuilderExtensions.cs]: Why empty? Do we need a comment explaining it?

Comment (jkotalik) [reply src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyBuilderExtensions.cs]: Sorry redesign for this. I need to do a bit of work still!

Comment (muratg) [reply src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyBuilderExtensions.cs]: Oh, OK, sorry didn't notice the [WIP] in the title.

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.HttpsPolicy
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsExtensions.cs]: Builder namespace

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _next = next;
+
+            var hstsOptions = options.Value;
+            var includeSubdomains = hstsOptions.IncludeSubDomains ? IncludeSubDomains : StringSegment.Empty;
+            var preload = hstsOptions.Preload ? Preload : StringSegment.Empty;
+
+            _nameValueHeaderValue = new StringValues($"max-age={hstsOptions.MaxAge}{includeSubdomains}{preload}");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: _stsValue

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <returns></returns>
+        public async Task Invoke(HttpContext context)
+        {
+
+            if (context.Request.IsHttps)
+            {
+                context.Response.Headers[HeaderNames.StrictTransportSecurity] = _nameValueHeaderValue;
+            }
+            await _next(context);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: no async/await required

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+    public class HstsOptions
+    {
+        /// <summary>
+        /// Sets the max-age parameter of the Strict-Transport-Security header.
+        /// </summary>
+        /// <remarks>
+        /// Max-age is required; defaults to 0.
+        /// See: https://tools.ietf.org/html/rfc6797#section-6.1.1
+        /// </remarks>
+        public int MaxAge { get; set; } // TODO set this to a different default in dev/not dev?
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs]: 0 cannot be the default, that has a specific meaning in the spec. Ask the PMs for a specific value. 1 day? 1 week?
We may not need a separate value for dev as the templates won't have this enabled by default.

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace Microsoft.AspNetCore.HttpsPolicy
+{
+    /// <summary>
+    /// Options for the HttpsPolicyMiddleware middleware
+    /// </summary>
+    public class HttpsPolicyOptions
+    {
+        /// <summary>
+        /// Whether to use HTTP Strict-Transport-Security (HSTS) on all HTTPS requests.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs]: responses

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs (hunk 12 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs]: Yes RedirectStatusCode.

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+        /// </summary>
+        public bool SetHsts { get; set; }
+
+        /// <summary>
+        /// The status code to be used for Url Redirection
+        /// </summary>
+        public int StatusCode { get; set; } = StatusCodes.Status301MovedPermanently; // TODO throw ArgumentOutOfRangeException from UrlRewrite redirect rule?
+
+        /// <summary>
+        /// The TLS port to be added to the redirected URL.
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsPolicyOptions.cs]: The default port (443) will be used if not provided.

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+        [InlineData(302, null, 0, false, false, "max-age=0", "https://localhost/")]
+        [InlineData(301, 5050, 0, false, false, "max-age=0", "https://localhost:5050/")]
+        [InlineData(301, 443, 0, false, false, "max-age=0", "https://localhost/")]
+        [InlineData(301, 443, 0, true, false, "max-age=0; includeSubDomains", "https://localhost/")]
+        [InlineData(301, 443, 0, false, true, "max-age=0; preload", "https://localhost/")]
+        [InlineData(301, null, 0, true, true, "max-age=0; includeSubDomains; preload", "https://localhost/")]
+        [InlineData(302, 5050, 0, true, true, "max-age=0; includeSubDomains; preload", "https://localhost:5050/")]
+        public async Task SetOptions_AddHstsMiddlewareAndEnableInPolicy(int statusCode, int? tlsPort, int maxAge, bool includeSubDomains, bool preload, string expectedHstsHeader, string expectedUrl)
+        {
+
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs]: nit: remove empty line

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+                    {
+                        options.IncludeSubDomains = includeSubDomains;
+                        options.MaxAge = maxAge;
+                        options.Preload = preload;
+                    });
+                })
+                .Configure(app =>
+                {
+                    app.UseHttpsPolicy();
+                    app.UseHsts();
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs]: I thought UseHttpsPolicy would add Hsts if needed, why is this line here. Is that not the case?

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs [hunk 23]
```diff
@@ -0,0 +1,257 @@
...
+                });
+
+            var server = new TestServer(builder);
+            var client = server.CreateClient();
+
+            var request = new HttpRequestMessage(HttpMethod.Get, "");
+
+            var response = await client.SendAsync(request);
+
+            // By default, we will not redirect if we only use the Hsts Middleware.
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs]: This comment is confusing, I'm not sure what you are trying to convey here, but it seems out of context

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs (hunk 22 above)
Comment (jkotalik) [reply test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyMiddlewareTests.cs]: That is the case. This is just showing that you can still add Hsts again after calling UseHttps.

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+                });
+
+            var server = new TestServer(builder);
+            var client = server.CreateClient();
+            client.BaseAddress = new Uri("https://localhost:5050");
+            var request = new HttpRequestMessage(HttpMethod.Get, "");
+
+            var response = await client.SendAsync(request);
+
+            // By default, we will not redirect if we only use the Hsts Middleware.
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs]: You copy pasted this comment below and doesn't make sense there. I don't think this adds a lot of value here, so I would just remove it.

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+        public async Task SetOptions_DefaultsSetCorrectly()
+        {
+            var builder = new WebHostBuilder()
+                .UseUrls("https://*:5050")
+                .ConfigureServices(services =>
+                {
+                })
+                .Configure(app =>
+                {
+                    app.UseHsts();
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs]: This is error prone as 0 is not a valid default

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs (hunk 18 above)
Comment (Tratcher) [reply src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs]: Discussed offline with @javiercn. 0 isn't a valid default, and using a nullable would unnecessarily complicate the user-experience. We just need to come up with a good default. Take it up with @blowdart over email if needs be.

Comment (jkotalik) [reply src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs]: @shirhatti as I know you were opinionated on having 0 as the default.

**Comment by jkotalik** [comment]
We should consider making HttpsPolicy a shared options collection with RedirectToHttpsOptions and HstsOptions. To configure the HttpPolicy, if we require the max-age to be set, users will have to call services.Configure<HstsOptions>.

File: src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs (hunk 3 above)
Comment (vanillajonathan) [reply src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: The RFC has it "IncludeSubDomains" but one could argue the property should be named `IncludeSubdomains`, I don't know.

Comment (jkotalik) [reply src/Microsoft.AspNetCore.HttpsEnforcement/HstsOptions.cs]: I'll stick with the RFC here. 😃

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs [hunk 26]
```diff
@@ -14,13 +14,13 @@ public class HstsOptions
         /// Sets the max-age parameter of the Strict-Transport-Security header.
         /// </summary>
         /// <remarks>
-        /// Max-age is required; defaults to 0.
+        /// Max-age is required; defaults to one month (2592000 seconds).
         /// See: https://tools.ietf.org/html/rfc6797#section-6.1.1
         /// </remarks>
-        public int MaxAge { get; set; }
+        public int MaxAge { get; set; } = 2592000;
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsOptions.cs]: This should be a TimeSpan so people don't have to do the math themselves.

File: samples/HttpsPolicySample/Startup.cs [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+namespace HttpsSample
+{
+    public class Startup
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<HttpsRedirectionOptions>(options => {
+                options.RedirectStatusCode = 301;
```
Comment (vanillajonathan) [inline samples/HttpsPolicySample/Startup.cs]: Use `StatusCodes.Status301MovedPermanently`

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs [hunk 28]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        /// <summary>
+        /// Invoke the middleware.
+        /// </summary>
+        /// <param name="context">The <see cref="HttpContext"/>.</param>
+        /// <returns></returns>
+        public Task Invoke(HttpContext context)
+        {
+
```
Comment (vanillajonathan) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: We've got a blank line here.

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs [hunk 29]
```diff
@@ -0,0 +1,66 @@
...
+            if (context.Request.IsHttps)
+            {
+                context.Response.Headers[HeaderNames.StrictTransportSecurity] = _strictTransportSecurityValue;
+            }
+            return  _next(context);
```
Comment (vanillajonathan) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: Should have a blank line before the return statement.

File: samples/HttpsPolicySample/Startup.cs [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace HttpsSample
+{
+    public class Startup
+    {
+        // This method gets called by the runtime. Use this method to add services to the container.
+        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.Configure<HttpsRedirectionOptions>(options => {
```
Comment (javiercn) [inline samples/HttpsPolicySample/Startup.cs]: Update the samples to use the helper methods

**Comment by jkotalik** [comment]
@javiercn @Tratcher We need to talk about how configuration options will be propagated to the middleware. Aka does this middleware need to do anything different or will configuration call app.UseHsts() or app.HttpsRedirection().

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs [hunk 31]
```diff
@@ -0,0 +1,66 @@
...
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            _next = next;
+
+            var hstsOptions = options.Value;
+            var includeSubdomains = hstsOptions.IncludeSubDomains ? IncludeSubDomains : StringSegment.Empty;
+            var preload = hstsOptions.Preload ? Preload : StringSegment.Empty;
+
+            _strictTransportSecurityValue = new StringValues($"max-age={hstsOptions.MaxAge.TotalSeconds}{includeSubdomains}{preload}");
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: TotalSeconds needs formatting. Compare to https://github.com/aspnet/Security/blob/bd07f8b683ce793490d108b2310fa6112953d172/src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectHandler.cs#L361-L362

File: src/Microsoft.AspNetCore.HttpsPolicy/HttpsRedirectionOptions.cs [hunk 32]
```diff
@@ -0,0 +1,27 @@
...
+using Microsoft.AspNetCore.Http;
+
+namespace Microsoft.AspNetCore.HttpsPolicy
+{
+    /// <summary>
+    /// Options for the HttpsRedirection middleware
+    /// </summary>
+    public class HttpsRedirectionOptions
+    {
+
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/HttpsRedirectionOptions.cs]: extra line

File: src/Microsoft.AspNetCore.HttpsPolicy/Microsoft.AspNetCore.HttpsPolicy.csproj [hunk 33]
```diff
@@ -0,0 +1,19 @@
...
+    <Description>
+      ASP.NET Core basic middleware for supporting HTTPS Redirection and HTTP Strict-Transport-Security.
+    </Description>
+    <TargetFramework>netstandard2.0</TargetFramework>
+    <NoWarn>$(NoWarn);CS1591</NoWarn>
+    <GenerateDocumentationFile>true</GenerateDocumentationFile>
+    <PackageTags>aspnetcore;https;hsts</PackageTags>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.AspNetCore.Http.Extensions" />
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.HttpsPolicy/Microsoft.AspNetCore.HttpsPolicy.csproj]: What is Http.Extensions needed for?

**Comment by Tratcher** [comment]
The current Options usage looks fine.

File: src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs (hunk 31 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: I also need to check for if the MaxAge has value then. Default to the same default in HstsOptions if no value?

Comment (jkotalik) [reply src/Microsoft.AspNetCore.HttpsPolicy/HstsMiddleware.cs]: nvrm realized I realized that security has max age as nullable.

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs [hunk 34]
```diff
@@ -0,0 +1,128 @@
...
+        [InlineData(50000, false, true, "max-age=50000; preload")]
+        [InlineData(0, true, true, "max-age=0; includeSubDomains; preload")]
+        [InlineData(50000, true, true, "max-age=50000; includeSubDomains; preload")]
+        public async Task SetOptionsThroughConfigure_SetsHeaderCorrectly(int maxAge, bool includeSubDomains, bool preload, string expected)
+        {
+            var builder = new WebHostBuilder()
+                .UseUrls("https://*:5050")
+                .ConfigureServices(services =>
+                {
+                    services.Configure<HstsOptions>(options => {
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs]: Fix the usages on the tests to use the service collection extensions

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyTests.cs [hunk 35]
```diff
@@ -0,0 +1,76 @@
...
+        [InlineData(301, null, 2592000, true, true, "max-age=2592000; includeSubDomains; preload", "https://localhost/")]
+        [InlineData(302, 5050, 2592000, true, true, "max-age=2592000; includeSubDomains; preload", "https://localhost:5050/")]
+        public async Task SetsBothHstsAndHttpsRedirection_RedirectOnFirstRequest_HstsOnSecondRequest(int statusCode, int? tlsPort, int maxAge, bool includeSubDomains, bool preload, string expectedHstsHeader, string expectedUrl)
+        {
+
+            var builder = new WebHostBuilder()
+                .UseUrls("https://*:5050", "http://*:5050")
+                .ConfigureServices(services =>
+                {
+                    services.Configure<HttpsRedirectionOptions>(options =>
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsPolicyTests.cs]: Here too

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsRedirectionMiddlewareTests.cs [hunk 36]
```diff
@@ -0,0 +1,125 @@
...
+        [InlineData(308, null, "https://localhost/")]
+        [InlineData(301, 5050, "https://localhost:5050/")]
+        [InlineData(301, 443, "https://localhost/")]
+        public async Task SetOptions_SetStatusCodeTlsPort(int statusCode, int? tlsPort, string expected)
+        {
+
+            var builder = new WebHostBuilder()
+                .ConfigureServices(services =>
+                {
+                    services.Configure<HttpsRedirectionOptions>(options =>
```
Comment (javiercn) [inline test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HttpsRedirectionMiddlewareTests.cs]: Here

File: test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs (hunk 34 above)
Comment (jkotalik) [reply test/Microsoft.AspNetCore.HttpsEnforcement.Tests/HstsMiddlewareTests.cs]: There are two sets of tests, one that uses Configure<HstsOptions> and AddHsts(). I can remove these tests if you would like 😄
